pylearn2/datasets/hdf5.py

```diff
@@ -2,338 +2,472 @@
-    convert : list, optional
-        A list of callables (in the same order as the sources in
-        data_specs) that will be applied to each slice of the dataset.
-    """
+        if isinstance(source_or_alias, tuple):
+            space = tuple([self.spaces[s] for s in source_or_alias])
+            space = CompositeSpace(space)
+        else:
+            space = self.spaces[source_or_alias]
+        return (space, source_or_alias)
 
-    def next(self):
+    #@wraps(Dataset.get, assigned=(), updated=())
+    # missing in dataset!
+    def get(self, sources, indexes):
         """
-        Get the next subset of the dataset during dataset iteration.
-
-        Converts index selections for batches to boolean selections that
-        are supported by HDF5 datasets.
+        Retrieves the requested elements from the dataset.
+
+        Parameter
+        ---------
+        sources : tuple
+            A tuple of source identifiers
+        indexes : slice or list
+            A slice or a list of indexes
+
+        Return
+        ------
+        rval : tuple
+            A tuple of batches, one for each source
         """
-        next_index = self._subset_iterator.next()
-
-        # convert to boolean selection
-        sel = np.zeros(self.num_examples, dtype=bool)
-        sel[next_index] = True
-        next_index = sel
+        assert isinstance(sources, tuple) and len(sources) > 0, (
+            'sources should be an instance of tuple and not empty')
+        assert all([isinstance(el, basestring) for el in sources]), (
+            'sources elements should be strings')
+        assert isinstance(indexes, (tuple, slice)), (
+            'indexes should be either a slice or a tuple of ints elements '
+            'should be strings')
+        if isinstance(indexes, tuple):
+            assert len(indexes) > 0 and all([isinstance(i, py_integer_types)
+                                            for i in indexes]), (
+                'indexes elements should be ints')
 
         rval = []
-        for data, fn in safe_izip(self._raw_data, self._convert):
+        for s in sources:
             try:
-                this_data = data[next_index]
-            except TypeError:
-                this_data = data[next_index, :]
-            if fn:
-                this_data = fn(this_data)
-            assert not contains_nan(this_data)
-            rval.append(this_data)
-        rval = tuple(rval)
-        if not self._return_tuple and len(rval) == 1:
-            rval, = rval
-        return rval
+                sdata = self.data[s]
+            except ValueError as e:
+                reraise_as(ValueError(
+                    'The requested source %s is not part of the dataset' %
+                    sources[s], *e.args))
+            if isinstance(indexes, slice) or len(indexes) == 1:
+                rval.append(sdata[indexes])
+            else:
+                warnings.warn('Accessing non sequential elements of an '
```

Comment:
Yes you are right. This way of accessing the data should really be avoided, I intended it more for single values access (for instance to check something) than for batch iteration, but still I agree that it is better to avoid the warnings flood. 

What if I add a flag to stop warning after the first 10 iterations?